Python/optimizer.c

```diff
@@ -935,3 +963,209 @@ PyUnstable_Optimizer_NewUOpOptimizer(void)
     opt->backedge_threshold = 16 << OPTIMIZER_BITS_IN_COUNTER;
     return (PyObject *)opt;
 }
+
+
+/*****************************************
+ *        Executor management
+ ****************************************/
+
+/* We use a bloomfilter with k = 6, m = 256
+ * The choice of k and the following constants
+ * could do with a more rigourous analysis,
+ * but here is a simple analysis:
+ *
+ * We want to keep the false positive rate low.
+ * For n = 5 (a trace depends on 5 objects),
+ * we expect 30 bits set, giving a false positive
+ * rate of (30/256)**6 == 2.5e-6 which is plenty
+ * good enough.
+ *
+ * However with n = 10 we expect 60 bits set (worst case),
+ * giving a false positive of (60/256)**6 == 0.0001
+ *
+ * We choose k = 6, rather than a higher number as
+ * it means the false positive rate grows slower for high n.
+ *
+ * n = 5, k = 6 => fp = 2.6e-6
+ * n = 5, k = 8 => fp = 3.5e-7
+ * n = 10, k = 6 => fp = 1.6e-4
+ * n = 10, k = 8 => fp = 0.9e-4
+ * n = 15, k = 6 => fp = 0.18%
+ * n = 15, k = 8 => fp = 0.23%
+ * n = 20, k = 6 => fp = 1.1%
+ * n = 20, k = 8 => fp = 2.3%
+ *
+ * The above analysis assumes perfect hash functions,
+ * but those don't exist, so the real false positive
+ * rates may be worse.
+ */
+
+#define K 6
+
+/* TO DO -- Use more modern hash functions with better distribution of bits */
+static uint32_t
+address_to_hash(void *ptr, uint32_t seed, uint32_t multiplier) {
+    assert(ptr != NULL);
+    uint32_t uhash = seed;
+    uintptr_t addr = (uintptr_t)ptr;
+    for (int i = 0; i < SIZEOF_VOID_P; i++) {
+        uhash ^= addr & 255;
+        uhash *= multiplier;
+        addr >>= 8;
+    }
+    return uhash;
+}
+
+static const uint32_t multipliers[2] = {
+    _PyHASH_MULTIPLIER,
+    110351524,
+};
+
+static const uint32_t seeds[2] = {
+    20221211,
+    2147483647,
+};
+
+void
+_Py_BloomFilter_Init(_PyBloomFilter *bloom)
+{
+    for (int i = 0; i < BLOOM_FILTER_WORDS; i++) {
+        bloom->bits[i] = 0;
+    }
+}
+
+void
+_Py_BloomFilter_Add(_PyBloomFilter *bloom, void *ptr)
+{
+    /* Set k (6) bits */
+    for (int i = 0; i < 2; i++) {
+        uint32_t hash = address_to_hash(ptr, seeds[i], multipliers[i]);
+        for (int j = 0; j < (K/2); j++) {
+            uint8_t bits = hash & 255;
+            bloom->bits[bits >> 5] |= (1 << (bits&31));
+            hash >>= 8;
+        }
```

Comment:
Here we learn that everything hangs together -- we're using a 32-bit hash function so we can use at most 4 bytes for a 256-bit bloom filter. In fact for K=6 we throw away the top 8 bits of the hash.

Reply:
I'll add a comment.